state/applicationoffers_test.go

```diff
@@ -500,16 +517,79 @@ func (s *applicationOffersSuite) addOfferConnection(c *gc.C, offerUUID string) {
 		Username:        "admin",
 		SourceModelUUID: testing.ModelTag.Id(),
 	})
+
+	return app
+}
+
+func (s *applicationOffersSuite) TestRemoveOffersSucceedsWithZeroConnections(c *gc.C) {
+	s.createDefaultOffer(c)
+	ao := state.NewApplicationOffers(s.State)
+	err := ao.Remove("hosted-mysql", false)
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.mysql.Destroy()
+	c.Assert(err, jc.ErrorIsNil)
 }
 
-func (s *applicationOffersSuite) TestRemoveOffersWithConnections(c *gc.C) {
+func (s *applicationOffersSuite) TestRemoveOffersFailsWithConnections(c *gc.C) {
 	offer := s.createDefaultOffer(c)
 	s.addOfferConnection(c, offer.OfferUUID)
 	ao := state.NewApplicationOffers(s.State)
 	err := ao.Remove("hosted-mysql", false)
 	c.Assert(err, gc.ErrorMatches, `cannot delete application offer "hosted-mysql": offer has 1 relation`)
 }
 
+func (s *applicationOffersSuite) TestRemoveOffersFailsWhenLateConnectionAppears(c *gc.C) {
+	offer := s.createDefaultOffer(c)
+	ao := state.NewApplicationOffers(s.State)
+	addOfferConnection := func() {
+		c.Logf("adding connection to %s", offer.OfferUUID)
+		s.addOfferConnection(c, offer.OfferUUID)
+	}
+	defer state.SetBeforeHooks(c, s.State, addOfferConnection).Check()
+
+	err := ao.Remove("hosted-mysql", false)
+	c.Assert(err, gc.ErrorMatches, `cannot delete application offer "hosted-mysql": offer has 1 relation`)
+}
+
+func (s *applicationOffersSuite) TestRemoveOffersSucceedsWhenLocalRelationAdded(c *gc.C) {
+	offer := s.createDefaultOffer(c)
+	s.AddTestingApplication(c, "local-wordpress", s.AddTestingCharm(c, "wordpress"))
+	app, err := s.State.Application(offer.ApplicationName)
+
+	addRelation := func() {
+		eps, err := s.State.InferEndpoints("local-wordpress", "mysql")
+		c.Assert(err, jc.ErrorIsNil)
+		_, err = s.State.AddRelation(eps...)
+		c.Assert(err, jc.ErrorIsNil)
+	}
+	defer state.SetBeforeHooks(c, s.State, addRelation).Check()
+
+	ao := state.NewApplicationOffers(s.State)
+	err = ao.Remove(offer.OfferName, false)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = app.Destroy()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *applicationOffersSuite) TestForceRemoveApplicationsWhenOffersHaveConnections(c *gc.C) {
```

Comment:
isn't this the same as TestForceRemoveApplicationsWithOffersThatHaveConnections

Reply:
yes, it was copied from there.. sorry about the churn.